source/common/router/header_parser.cc

```diff
@@ -351,19 +358,34 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
       switch (entry.append_action_) {
         PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
       case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
-        headers.addReferenceKey(key, value);
+        headers_to_add.emplace_back(key, value);
```

Comment:
perf nit: if stream_info is nullptr then a pointless string-copy happens for every header.

Consider having the stream_info!=null check outside the loop and do something entirely different in that case.

Reply:
@jmarantz Would you mind elaborating more about:

`perf nit: if stream_info is nullptr then a pointless string-copy happens for every header.` ? Thanks!

Reply:
it looks like it's valid to call evaluateHeaders with stream_info==nullptr. If you follow the code, you'll. see we check for that for every header in headers_to_add_, and we'll be executing
```
  const std::string value = false ? .... : entry.original_value;
```
that will do a string copy and allocation which is not needed. Instead you could hoist the `stream_info!=nullptr` check outside the loop (saving a bunch of branches even if stream_info!=nullptr) and early-exit (saving allocations and string copies if stream_info==nullptr).

Reply:
Got it. Thanks for explaining. Let me try to refactor it and benchmark again.

Reply:
Going back to string copy: I thought that `std::string`'s copy constructor and assignment do not really copy the underlying string, but only where the string is. The copy happens only when one of the instances decides to modify the string and produces its own copy of underlying string before modifying it. Not sure if this is still the case, but I remember reading about that some time ago.

Reply:
That was true 10+ years ago (copy-on-write) but is no longer true at least for the STL impls I know. The locking overhead needed for that feature turned out to make it a net loss.